File: backend/app/api/v1/endpoints/campaigns.py
from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from sqlalchemy.orm import Session
from typing import List, Optional
from uuid import UUID

from app.core.database import get_db
from app.models.user import User
from app.services.campaign_service import CampaignAnalyzerService
from app.services.auth_service import AuthService
from app.core.security import verify_token
from app.services.logging_service import get_logging_service, LogModule
from app.schemas.campaign import (
    Campaign, CampaignCreate, CampaignUpdate, CampaignListResponse,
    CampaignQuestionnaire, StartAnalysisRequest,
    CampaignAnalysis, CampaignAnalysisResponse
)

router = APIRouter()
security = HTTPBearer()
logging_service = get_logging_service()

# Constants
CAMPAIGN_NOT_FOUND = "Campaign not found"

# Dependency to get current user with proper JWT authentication
async def get_current_user(
    credentials: HTTPAuthorizationCredentials = Depends(security),
    db: Session = Depends(get_db)
) -> User:
    """Get current user from JWT token."""
    try:
        payload = verify_token(credentials.credentials)
        user = AuthService.get_user_by_email(db, payload.get("sub"))
        if not user:
            raise HTTPException(
                status_code=401,
                detail="User not found"
            )
        return user
    except Exception as e:
        logging_service.log(LogModule.AUTH, f"Invalid authentication credentials: {e}", "ERROR")
        raise HTTPException(
            status_code=401,
            detail="Invalid authentication credentials"
        )

@router.post("/", response_model=Campaign)
async def create_campaign(
    campaign_data: CampaignCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create a new campaign"""
    service = CampaignAnalyzerService(db)
    campaign = service.create_campaign(campaign_data, current_user.id, current_user.org_id)
    return campaign

@router.get("/", response_model=CampaignListResponse)
async def get_campaigns(
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    status: Optional[str] = Query(None),
    campaign_type: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get campaigns with filtering and pagination"""
    service = CampaignAnalyzerService(db)
    campaigns, total = service.get_campaigns(
        org_id=current_user.org_id,
        user_id=None,  # Get all campaigns for the organization
        status=status,
        campaign_type=campaign_type,
        page=page,
        per_page=per_page
    )
    
    return CampaignListResponse(
        campaigns=campaigns,
        total=total,
        page=page,
        per_page=per_page
    )

@router.get("/{campaign_id}", response_model=Campaign)
async def get_campaign(
    campaign_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get a specific campaign"""
    service = CampaignAnalyzerService(db)
    campaign = service.get_campaign(campaign_id, current_user.org_id)
    
    if not campaign:
        raise HTTPException(status_code=404, detail=CAMPAIGN_NOT_FOUND)
    
    return campaign

@router.put("/{campaign_id}", response_model=Campaign)
async def update_campaign(
    campaign_id: UUID,
    campaign_data: CampaignUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update a campaign"""
    service = CampaignAnalyzerService(db)
    campaign = service.update_campaign(campaign_id, campaign_data, current_user.org_id)
    
    if not campaign:
        raise HTTPException(status_code=404, detail=CAMPAIGN_NOT_FOUND)
    
    return campaign

@router.delete("/{campaign_id}")
async def delete_campaign(
    campaign_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Delete a campaign"""
    service = CampaignAnalyzerService(db)
    success = service.delete_campaign(campaign_id, current_user.org_id)
    
    if not success:
        raise HTTPException(status_code=404, detail=CAMPAIGN_NOT_FOUND)
    
    return {"message": "Campaign deleted successfully"}

@router.post("/from-questionnaire", response_model=Campaign)
async def create_campaign_from_questionnaire(
    questionnaire: CampaignQuestionnaire,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create a campaign from questionnaire data"""
    service = CampaignAnalyzerService(db)
    campaign = service.create_campaign_from_questionnaire(
        questionnaire, current_user.id, current_user.org_id
    )
    return campaign

@router.post("/{campaign_id}/analyze", response_model=CampaignAnalysis)
async def start_campaign_analysis(
    campaign_id: UUID,
    request: StartAnalysisRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Start campaign analysis"""
    service = CampaignAnalyzerService(db)
    
    # Check if campaign exists
    campaign = service.get_campaign(campaign_id, current_user.org_id)
    if not campaign:
        raise HTTPException(status_code=404, detail=CAMPAIGN_NOT_FOUND)
    
    # Update campaign with questionnaire data if provided
    if request.questionnaire_data:
        # You might want to update the campaign with new questionnaire data
        pass
    
    analysis = service.start_campaign_analysis(
        campaign_id, current_user.id, current_user.org_id, request.force_reanalysis
    )
    
    return analysis

@router.post("/analyze-questionnaire", response_model=CampaignAnalysisResponse)
async def analyze_questionnaire(
    request: StartAnalysisRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Create campaign from questionnaire and start analysis"""
    service = CampaignAnalyzerService(db)
    
    # Create campaign from questionnaire
    campaign = service.create_campaign_from_questionnaire(
        request.questionnaire_data, current_user.id, current_user.org_id
    )
    
    # Start analysis
    analysis = service.start_campaign_analysis(
        campaign.id, current_user.id, current_user.org_id, True
    )
    
    return CampaignAnalysisResponse(
        analysis=analysis,
        campaign=campaign
    )

@router.get("/{campaign_id}/analysis", response_model=CampaignAnalysis)
async def get_campaign_analysis(
    campaign_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get the latest analysis for a campaign"""
    service = CampaignAnalyzerService(db)
    
    # Check if campaign exists
    campaign = service.get_campaign(campaign_id, current_user.org_id)
    if not campaign:
        raise HTTPException(status_code=404, detail=CAMPAIGN_NOT_FOUND)
    
    analysis = service.get_campaign_analysis(campaign_id, current_user.org_id)
    
    if not analysis:
        raise HTTPException(status_code=404, detail="No analysis found for this campaign")
    
    return analysis

@router.get("/{campaign_id}/analysis/full", response_model=CampaignAnalysisResponse)
async def get_campaign_analysis_full(
    campaign_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get campaign analysis with full campaign details"""
    service = CampaignAnalyzerService(db)
    
    # Check if campaign exists
    campaign = service.get_campaign(campaign_id, current_user.org_id)
    if not campaign:
        raise HTTPException(status_code=404, detail=CAMPAIGN_NOT_FOUND)
    
    analysis = service.get_campaign_analysis(campaign_id, current_user.org_id)
    
    if not analysis:
        raise HTTPException(status_code=404, detail="No analysis found for this campaign")
    
    return CampaignAnalysisResponse(
        analysis=analysis,
        campaign=campaign
    )

# Additional utility endpoints
@router.get("/types/options")
async def get_campaign_type_options():
    """Get available campaign type options"""
    return {
        "campaign_types": [
            {"value": "search", "label": "Search", "description": "Text ads on search results"},
            {"value": "display", "label": "Display", "description": "Visual ads on websites"},
            {"value": "video", "label": "Video", "description": "Video ads on platforms like YouTube"},
            {"value": "shopping", "label": "Shopping", "description": "Product ads with images and prices"},
            {"value": "performance_max", "label": "Performance Max", "description": "AI-driven cross-platform campaigns"},
            {"value": "app", "label": "App", "description": "Promote mobile app installs"},
            {"value": "local", "label": "Local", "description": "Drive visits to physical locations"}
        ],
        "platforms": [
            {"value": "google_ads", "label": "Google Ads", "description": "Google's advertising platform"},
            {"value": "facebook", "label": "Facebook", "description": "Facebook advertising"},
            {"value": "instagram", "label": "Instagram", "description": "Instagram advertising"},
            {"value": "linkedin", "label": "LinkedIn", "description": "Professional network advertising"},
            {"value": "twitter", "label": "Twitter", "description": "Twitter advertising"},
            {"value": "tiktok", "label": "TikTok", "description": "TikTok advertising"},
            {"value": "snapchat", "label": "Snapchat", "description": "Snapchat advertising"},
            {"value": "pinterest", "label": "Pinterest", "description": "Pinterest advertising"}
        ],
        "objectives": [
            {"value": "awareness", "label": "Brand Awareness", "description": "Increase brand visibility"},
            {"value": "traffic", "label": "Website Traffic", "description": "Drive visitors to website"},
            {"value": "leads", "label": "Lead Generation", "description": "Generate leads and inquiries"},
            {"value": "sales", "label": "Sales", "description": "Drive online sales"},
            {"value": "conversions", "label": "Conversions", "description": "Drive specific actions"},
            {"value": "engagement", "label": "Engagement", "description": "Increase social engagement"},
            {"value": "app_installs", "label": "App Installs", "description": "Drive mobile app downloads"}
        ]
    }

@router.get("/templates/questionnaire")
async def get_questionnaire_template():
    """Get questionnaire template for frontend"""
    return {
        "sections": [
            {
                "id": "basic_info",
                "title": "Basic Information",
                "fields": [
                    {
                        "name": "campaign_name",
                        "type": "text",
                        "label": "Campaign Name",
                        "required": True,
                        "placeholder": "Enter a descriptive name for your campaign"
                    },
                    {
                        "name": "business_goals",
                        "type": "multiselect",
                        "label": "Primary Business Goals",
                        "required": True,
                        "options": ["Brand Awareness", "Website Traffic", "Lead Generation", "Sales", "Engagement"]
                    },
                    {
                        "name": "target_action",
                        "type": "text",
                        "label": "What action do you want users to take?",
                        "required": True,
                        "placeholder": "e.g., Purchase product, Sign up for newsletter, Download app"
                    }
                ]
            },
            {
                "id": "budget_timeline",
                "title": "Budget & Timeline",
                "fields": [
                    {
                        "name": "budget_range",
                        "type": "select",
                        "label": "Budget Range",
                        "required": True,
                        "options": ["$500-1K", "$1K-5K", "$5K-10K", "$10K-25K", "$25K+"]
                    },
                    {
                        "name": "campaign_duration",
                        "type": "select",
                        "label": "Campaign Duration",
                        "required": True,
                        "options": ["1-2 weeks", "1 month", "2-3 months", "6 months", "Ongoing"]
                    },
                    {
                        "name": "seasonality",
                        "type": "text",
                        "label": "Seasonal Considerations",
                        "required": True,
                        "placeholder": "Are there seasonal factors that affect your business?"
                    }
                ]
            },
            {
                "id": "audience",
                "title": "Target Audience",
                "fields": [
                    {
                        "name": "target_audience_description",
                        "type": "textarea",
                        "label": "Describe Your Target Audience",
                        "required": True,
                        "placeholder": "Who is your ideal customer? Be as specific as possible."
                    },
                    {
                        "name": "audience_interests",
                        "type": "tags",
                        "label": "Audience Interests",
                        "required": False,
                        "placeholder": "Enter interests and hobbies of your target audience"
                    },
                    {
                        "name": "audience_pain_points",
                        "type": "tags",
                        "label": "Audience Pain Points",
                        "required": False,
                        "placeholder": "What problems does your audience face?"
                    }
                ]
            }
        ]
    }

File: frontend/lib/campaigns.ts
/**
 * Campaign API service for frontend
 */

// Types
export interface Campaign {
  id: string
  name: string
  description?: string
  campaign_type: 'search' | 'display' | 'video' | 'shopping' | 'performance_max' | 'app' | 'local'
  platform: 'google_ads' | 'facebook' | 'instagram' | 'linkedin' | 'twitter' | 'tiktok' | 'snapchat' | 'pinterest'
  status: 'draft' | 'active' | 'paused' | 'completed'
  primary_objective: string
  secondary_objectives: string[]
  target_kpis: Record<string, any>
  total_budget?: number
  daily_budget?: number
  currency: string
  start_date?: string
  end_date?: string
  target_demographics: Record<string, any>
  target_locations: string[]
  target_interests: string[]
  target_behaviors: string[]
  audience_size_estimate?: number
  creative_assets: Record<string, any>
  messaging_themes: string[]
  call_to_action?: string
  landing_page_url?: string
  bidding_strategy?: string
  ad_scheduling: Record<string, any>
  device_targeting: string[]
  tags: string[]
  custom_fields: Record<string, any>
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface CampaignCreate {
  name: string
  description?: string
  campaign_type: string
  platform: string
  primary_objective: string
  secondary_objectives?: string[]
  target_kpis?: Record<string, any>
  total_budget?: number
  daily_budget?: number
  currency?: string
  start_date?: string
  end_date?: string
  target_demographics?: Record<string, any>
  target_locations?: string[]
  target_interests?: string[]
  target_behaviors?: string[]
  creative_assets?: Record<string, any>
  messaging_themes?: string[]
  call_to_action?: string
  landing_page_url?: string
  bidding_strategy?: string
  ad_scheduling?: Record<string, any>
  device_targeting?: string[]
  tags?: string[]
  custom_fields?: Record<string, any>
}

export interface CampaignListResponse {
  campaigns: Campaign[]
  total: number
  page: number
  per_page: number
}

export interface CampaignQuestionnaire {
  campaign_name: string
  business_goals: string[]
  target_action: string
  budget_range: string
  campaign_duration: string
  seasonality: string
  target_audience_description: string
  audience_interests?: string[]
  audience_pain_points?: string[]
}

// Campaign Designer types
export interface CampaignDesignerData {
  name: string
  objective: 'lead_gen' | 'ecommerce_sales' | 'app_installs' | 'awareness'
  primaryKpi: 'CPL' | 'CPA' | 'ROAS' | 'CTR' | 'Reach'
  budget: {
    total: number
    daily: number
    duration: number
  }
  channels: Array<{
    channel: string
    percentage: number
    amount: number
  }>
  targeting: {
    geo: string[]
    audience: string
  }
  kpiTarget: number
  designScore: number
  selectedTemplate?: string
  createdAt: string
}

export interface CampaignDesignerCreate {
  designer_data: CampaignDesignerData
}

export interface CampaignAnalysis {
  id: string
  campaign_id: string
  analysis_type: string
  status: 'pending' | 'processing' | 'completed' | 'failed'
  ga4_data: Record<string, any>
  campaign_predictions: Record<string, any>
  performance_gaps: Record<string, any>
  recommendations: Record<string, any>
  priority_actions: string[]
  overall_score?: number
  gap_scores: Record<string, any>
  confidence_level?: number
  audience_analysis: Record<string, any>
  creative_analysis: Record<string, any>
  budget_analysis: Record<string, any>
  timing_analysis: Record<string, any>
  technical_analysis: Record<string, any>
  market_insights: Record<string, any>
  competitor_analysis: Record<string, any>
  seasonal_trends: Record<string, any>
  processing_started_at?: string
  processing_completed_at?: string
  error_message?: string
  created_at: string
  updated_at: string
}

export interface CampaignOptions {
  campaign_types: Array<{
    value: string
    label: string
    description: string
  }>
  platforms: Array<{
    value: string
    label: string
    description: string
  }>
  objectives: Array<{
    value: string
    label: string
    description: string
  }>
}

// API Configuration
const API_BASE_URL = process.env.NODE_ENV === 'production' 
  ? 'https://api-staging.nexopeak.ca'
  : 'http://localhost:8000'

/**
 * Get authentication headers
 */
function getAuthHeaders(): Record<string, string> {
  const token = localStorage.getItem('access_token')
  return {
    'Content-Type': 'application/json',
    ...(token && { 'Authorization': `Bearer ${token}` })
  }
}

/**
 * Handle API response
 */
async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const error = await response.text()
    throw new Error(`API Error: ${response.status} - ${error}`)
  }
  return response.json()
}

/**
 * Get campaign type options
 */
export async function getCampaignOptions(): Promise<CampaignOptions> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/campaigns/types/options`, {
      method: 'GET',
      headers: getAuthHeaders()
    })
    return handleResponse<CampaignOptions>(response)
  } catch (error) {
    console.error('Error fetching campaign options:', error)
    // Fallback data
    return {
      campaign_types: [
        { value: 'search', label: 'Search', description: 'Text ads on search results' },
        { value: 'display', label: 'Display', description: 'Visual ads on websites' },
        { value: 'video', label: 'Video', description: 'Video ads on platforms like YouTube' },
        { value: 'social', label: 'Social Media', description: 'Social media advertising' }
      ],
      platforms: [
        { value: 'google_ads', label: 'Google Ads', description: 'Google advertising platform' },
        { value: 'facebook', label: 'Facebook', description: 'Facebook advertising' },
        { value: 'instagram', label: 'Instagram', description: 'Instagram advertising' }
      ],
      objectives: [
        { value: 'awareness', label: 'Brand Awareness', description: 'Increase brand visibility' },
        { value: 'traffic', label: 'Website Traffic', description: 'Drive visitors to website' },
        { value: 'leads', label: 'Lead Generation', description: 'Generate leads and inquiries' },
        { value: 'sales', label: 'Sales', description: 'Drive online sales' }
      ]
    }
  }
}

/**
 * Get questionnaire template
 */
export async function getQuestionnaireTemplate(): Promise<any> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/campaigns/templates/questionnaire`, {
      method: 'GET',
      headers: getAuthHeaders()
    })
    return handleResponse(response)
  } catch (error) {
    console.error('Error fetching questionnaire template:', error)
    throw error
  }
}

/**
 * Get campaigns with pagination and filtering
 */
export async function getCampaigns(
  page: number = 1,
  per_page: number = 20,
  status?: string,
  campaign_type?: string
): Promise<CampaignListResponse> {
  try {
    const params = new URLSearchParams({
      page: page.toString(),
      per_page: per_page.toString(),
      ...(status && { status }),
      ...(campaign_type && { campaign_type })
    })

    const response = await fetch(`${API_BASE_URL}/api/v1/campaigns/?${params}`, {
      method: 'GET',
      headers: getAuthHeaders()
    })
    return handleResponse<CampaignListResponse>(response)
  } catch (error) {
    console.error('Error fetching campaigns:', error)
    // Return empty result on error
    return {
      campaigns: [],
      total: 0,
      page: 1,
      per_page: 20
    }
  }
}

/**
 * Get a specific campaign
 */
export async function getCampaign(campaignId: string): Promise<Campaign> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/campaigns/${campaignId}`, {
      method: 'GET',
      headers: getAuthHeaders()
    })
    return handleResponse<Campaign>(response)
  } catch (error) {
    console.error('Error fetching campaign:', error)
    throw error
  }
}

/**
 * Create a new campaign
 */
export async function createCampaign(campaignData: CampaignCreate): Promise<Campaign> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/campaigns/`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify(campaignData)
    })
    return handleResponse<Campaign>(response)
  } catch (error) {
    console.error('Error creating campaign:', error)
    throw error
  }
}


/**
 * Update a campaign
 */
export async function updateCampaign(campaignId: string, campaignData: Partial<CampaignCreate>): Promise<Campaign> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/campaigns/${campaignId}`, {
      method: 'PUT',
      headers: getAuthHeaders(),
      body: JSON.stringify(campaignData)
    })
    return handleResponse<Campaign>(response)
  } catch (error) {
    console.error('Error updating campaign:', error)
    throw error
  }
}

/**
 * Delete a campaign
 */
export async function deleteCampaign(campaignId: string): Promise<void> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/campaigns/${campaignId}`, {
      method: 'DELETE',
      headers: getAuthHeaders()
    })
    await handleResponse(response)
  } catch (error) {
    console.error('Error deleting campaign:', error)
    throw error
  }
}

/**
 * Start campaign analysis
 */
export async function startCampaignAnalysis(
  campaignId: string, 
  questionnaireData?: CampaignQuestionnaire,
  forceReanalysis: boolean = false
): Promise<CampaignAnalysis> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/campaigns/${campaignId}/analyze`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify({
        questionnaire_data: questionnaireData,
        force_reanalysis: forceReanalysis
      })
    })
    return handleResponse<CampaignAnalysis>(response)
  } catch (error) {
    console.error('Error starting campaign analysis:', error)
    throw error
  }
}

/**
 * Analyze questionnaire (create campaign and start analysis)
 */
export async function analyzeQuestionnaire(questionnaire: CampaignQuestionnaire): Promise<{
  analysis: CampaignAnalysis
  campaign: Campaign
}> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/campaigns/analyze-questionnaire`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify({
        questionnaire_data: questionnaire,
        force_reanalysis: true
      })
    })
    return handleResponse(response)
  } catch (error) {
    console.error('Error analyzing questionnaire:', error)
    throw error
  }
}

/**
 * Get campaign analysis
 */
export async function getCampaignAnalysis(campaignId: string): Promise<CampaignAnalysis> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/campaigns/${campaignId}/analysis`, {
      method: 'GET',
      headers: getAuthHeaders()
    })
    return handleResponse<CampaignAnalysis>(response)
  } catch (error) {
    console.error('Error fetching campaign analysis:', error)
    throw error
  }
}

/**
 * Get full campaign analysis with campaign details
 */
export async function getCampaignAnalysisFull(campaignId: string): Promise<{
  analysis: CampaignAnalysis
  campaign: Campaign
}> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/campaigns/${campaignId}/analysis/full`, {
      method: 'GET',
      headers: getAuthHeaders()
    })
    return handleResponse(response)
  } catch (error) {
    console.error('Error fetching full campaign analysis:', error)
    throw error
  }
}

/**
 * Create campaign from Campaign Designer data
 */
export async function createCampaignFromDesigner(designerData: CampaignDesignerData): Promise<Campaign> {
  try {
    console.log('Sending campaign data to API:', designerData);
    console.log('API URL:', `${API_BASE_URL}/api/v1/campaigns/from-designer`);
    console.log('Headers:', getAuthHeaders());
    
    const response = await fetch(`${API_BASE_URL}/api/v1/campaigns/from-designer`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify({
        designer_data: designerData
      })
    })
    
    console.log('API Response status:', response.status);
    console.log('API Response headers:', response.headers);
    
    if (!response.ok) {
      const errorText = await response.text();
      console.error('API Error response:', errorText);
      throw new Error(`API Error: ${response.status} - ${errorText}`);
    }
    
    const result = await response.json();
    console.log('API Success response:', result);
    return result;
  } catch (error) {
    console.error('Error creating campaign from designer:', error)
    throw error
  }
}




